### DISCUSSION TITLE
Should we open-source Metalama.Compiler ?

### DISCUSSION THREAD

**Discussion by gfraiteur** [metalama/Metalama#174] [Ideas]
Metalama is a high-level meta-programming framework for C#. It uses most Roslyn extension points: analyzers, suppressors, source generators, code fix providers, and refactoring providers -- plus other VSX extension points.

However, one extension point is missing in Roslyn: _source transformers_, the ability to perform arbitrary transformations on the AST between parsing and compiling the code.  This is why we forked Roslyn into a project named [Metalama.Compiler](https://www.nuget.org/packages/Metalama.Compiler), which can be extended thanks to [Metalama.Compiler.Sdk](https://www.nuget.org/packages/Metalama.Compiler.Sdk). The only package that uses this extensibility point is [Metalama.Framework](https://www.nuget.org/packages/Metalama.Framework).

`Metalama.Compiler` is currently a closed-source project. We are considering contributing back `Metalama.Compiler ` to its original MIT license.

What do you guys think about this idea? Would you have any use of it?

**Comment by XtroTheArctic** [comment]
I wonder if anyone from Metalama/PostSharp community would be against Metalama.Compiler going open-source. I wouldn't.

**Comment by gfraiteur** [comment]
I'm interested to hear the potential motivations and use cases.
The first audience of this question is actually the Roslyn community. See https://github.com/dotnet/csharplang/issues/7009#issuecomment-1615050668

**Comment by WhitWaldo** [comment]
I, for one, am quite curious what it looks like, so I'd vote yes. At the same time though, I'm not entirely sure if I have a practical use case just yet of how else I'd use or extend the idea. Should I think of one though, I'd be happy to circle back to the thread and share it.

**Comment by PeterHevesi** [comment]
Yeah, definitely you should open source it :) I was talking to some guys in Roslyn community on Discord, and asked them what they think about Metalama, and they said that it looks very interesting, but they hated, that it's not free and open source :D 
So I think to open source the Metalama.Compiler would be very nice, especially when the Roslyn community wants to do interceptors, and I really believe that your expertise could be very helpful. I was reading that thread, and you should definitely show them your approach :)

**Comment by esbenbach** [comment]
I am rarely digging into the roslyn compiler, and never considered contributing, but every now and again I find myself looking at the code to understand some small detail, be that a compiler error or just a curiosity. That is only possible if the code is available.

Not sure that qualifies as sufficient insentive, but even for projects without any sizable communities or little to no interactions and contributions, it is still really beneficial to the consumers/users to see the code when there is something "weird" going on :)

**Comment by PeterHevesi** [comment]
Soo, will you open source it?

**Comment by gfraiteur** [comment]
I don't know. The idea was to get some interest from the Roslyn community but we didn't get any. The benefits for the Metalama community are very limited.

**Comment by PeterHevesi** [comment]
Well, you got some interest, that post in that Roslyn thread received few upvotes :)
So it's up to you, but I still think it would be a smart decision :)
